diff --git a/vowpalwabbit/gd.cc b/vowpalwabbit/gd.cc
index 623d6869..b3de60b7 100644
--- a/vowpalwabbit/gd.cc
+++ b/vowpalwabbit/gd.cc
@@ -468,7 +468,7 @@ inline void pred_per_update_feature(norm_data& nd, float x, float& fw)
       }
       nd.norm_x += x2 / (w[normalized] * w[normalized]);
     }
-    w[spare] = compute_rate_decay<sqrt_rate, adaptive, normalized>(nd.pd, w[0]);
+    w[spare] = compute_rate_decay<sqrt_rate, adaptive, normalized>(nd.pd, fw);
     nd.pred_per_update += x2 * w[spare];
   }
 }
